src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/services/StaticInvalidationService.kt

```diff
@@ -0,0 +1,156 @@
+     *
+     * @param fileUrl the url of a file
+     * @param lines the lines to invalidate tests
+     * @param cache the cache
+     */
+    fun invalidateCacheLines(fileUrl: String, lines: Set<Int>, cache: TestCaseCachingService) {
+        for (line in lines) {
+            cache.invalidateFromCache(fileUrl, line + 1, line + 1)
+        }
+    }
+
+    /**
+     * Checks for a file what lines have been modified
+     * @param file the file in question
+     * @return list of lines that have been modified
+     */
+    fun invalidate(file: PsiFile): Set<Int> {
+        // Initialize list
+        val linesToDiscard: MutableSet<Int> = mutableSetOf()
+        // Get basic details
+        val filePath = file.virtualFile.presentableUrl
+        val classToValidate = file.children.filterIsInstance<PsiClass>()
+        val doc: Document = PsiDocumentManager.getInstance(file.project).getDocument(file) ?: return setOf()
+        // Go over all classes in file
+        classToValidate.forEach { currentClass ->
+            val className = currentClass.name!!
+            val methods = currentClass.methods
+            val map: HashMap<String, Pair<ArrayList<PsiElement>, Set<Int>>> = HashMap()
+            methods.forEach {
+                val startLine = doc.getLineNumber(it.identifyingElement!!.startOffset)
+                val endLine = doc.getLineNumber(it.body!!.rBrace!!.endOffset)
+                map[it.hierarchicalMethodSignature.toString()] = Pair(recursePsiMethodBody(it.body!!), startLine.rangeTo(endLine).toSet())
+            }
+            // validate each class
+            linesToDiscard.addAll(validateClass(filePath, map, className))
+        }
+
+        return linesToDiscard
+    }
+
+    /**
+     * Returns a list of PsiElements that are part of the method psi children.
+     *
+     * @param psiMethodBody the psiBody of the method
+     * @return the list of PsiElements
+     */
+    private fun recursePsiMethodBody(psiMethodBody: PsiCodeBlock): ArrayList<PsiElement> {
+        val psiList: ArrayList<PsiElement> = arrayListOf()
+        for (psiStatement in psiMethodBody.statements) {
+            if (psiStatement.children.isEmpty()) {
+                psiList.add(psiStatement)
+                continue
+            }
+            for (psiElement in psiStatement.children) {
+                recurseTree(psiElement, psiList)
+            }
+        }
+        return psiList
+    }
+
+    /**
+     * Append elements to the list of PsiElements
+     *
+     * @param psiElement the psi of the element
+     * @param psiList list to append psi elements
+     */
+    private fun recurseTree(psiElement: PsiElement, psiList: ArrayList<PsiElement>) {
+        if (psiElement is PsiWhiteSpace || psiElement is PsiReferenceParameterList || psiElement is PsiModifierList) {
+            return
+        }
+        if (psiElement.children.isEmpty()) {
+            psiList.add(psiElement)
+            return
+        }
```

Comment:
Why do you add the element to the list only if it has no children?

Reply:
This is @kirilvasilev16 domain of knowledge. In general, this was just a way to recursively parse a method.

Reply:
This is necessary, because IntelliJ creates `PsiStatement` for every line (including those which only have whitespaces). For this reason, it is better to store only the `PsiElements` that contain no children in the snapshot, because they are actually the ones that make up the entire code inside a method/class.

Reply:
Oh, I see, that makes a lot of sense!